File: backend-node/src/services/accountService.js
/**
 * Account service for user management operations.
 */
const { User } = require('../models');
const { getPasswordHash, verifyPassword } = require('./authService');

/**
 * Create a new user account.
 * @param {object} userData - User data
 * @returns {Promise<[User|null, string|null]>} Tuple of (User, null) on success or (null, error_message) on failure
 */
async function createUser({ email, username, password, riotId = null, region = null }) {
  // Check if email already exists
  const existingEmail = await User.findOne({ where: { email } });
  if (existingEmail) {
    return [null, 'Email already registered'];
  }

  // Check if username already exists
  const existingUsername = await User.findOne({ where: { username } });
  if (existingUsername) {
    return [null, 'Username already taken'];
  }

  try {
    // Create new user
    const hashedPassword = await getPasswordHash(password);
    const user = await User.create({
      email,
      username,
      hashedPassword,
      riotId,
      region,
    });
    return [user, null];
  } catch (error) {
    console.error('Failed to create user:', error);
    return [null, 'Failed to create account'];
  }
}

/**
 * Authenticate a user by email and password.
 * @param {string} email - User email
 * @param {string} password - User password
 * @returns {Promise<User|null>} User if authentication successful, null otherwise
 */
async function authenticateUser(email, password) {
  const user = await User.findOne({ where: { email } });
  
  if (!user) {
    return null;
  }

  const passwordValid = await verifyPassword(password, user.hashedPassword);
  if (!passwordValid) {
    return null;
  }

  // Update last login
  user.lastLogin = new Date();
  await user.save();

  return user;
}

/**
 * Get a user by their ID.
 * @param {number} userId - User ID
 * @returns {Promise<User|null>}
 */
async function getUserById(userId) {
  return User.findByPk(userId);
}

/**
 * Get a user by their email.
 * @param {string} email - User email
 * @returns {Promise<User|null>}
 */
async function getUserByEmail(email) {
  return User.findOne({ where: { email } });
}

/**
 * Get a user by their username.
 * @param {string} username - Username
 * @returns {Promise<User|null>}
 */
async function getUserByUsername(username) {
  return User.findOne({ where: { username } });
}

/**
 * Update user profile information.
 * @param {number} userId - User ID
 * @param {object} updates - Fields to update
 * @returns {Promise<[User|null, string|null]>} Tuple of (User, null) on success or (null, error_message) on failure
 */
async function updateUser(userId, { username, riotId, region }) {
  const user = await User.findByPk(userId);
  
  if (!user) {
    return [null, 'User not found'];
  }

  // Check username uniqueness if changing
  if (username && username !== user.username) {
    const existing = await User.findOne({ where: { username } });
    if (existing) {
      return [null, 'Username already taken'];
    }
    user.username = username;
  }

  if (riotId !== undefined) {
    user.riotId = riotId;
  }

  if (region !== undefined) {
    user.region = region;
  }

  try {
    await user.save();
    return [user, null];
  } catch (error) {
    console.error('Failed to update user:', error);
    return [null, 'Failed to update profile'];
  }
}

/**
 * Change user's password.
 * @param {number} userId - User ID
 * @param {string} currentPassword - Current password
 * @param {string} newPassword - New password
 * @returns {Promise<[boolean, string|null]>} Tuple of (success, error_message)
 */
async function changePassword(userId, currentPassword, newPassword) {
  const user = await User.findByPk(userId);
  
  if (!user) {
    return [false, 'User not found'];
  }

  const passwordValid = await verifyPassword(currentPassword, user.hashedPassword);
  if (!passwordValid) {
    return [false, 'Current password is incorrect'];
  }

  try {
    user.hashedPassword = await getPasswordHash(newPassword);
    await user.save();
    return [true, null];
  } catch (error) {
    console.error('Failed to change password:', error);
    return [false, 'Failed to change password'];
  }
}

/**
 * Delete a user account.
 * @param {number} userId - User ID
 * @returns {Promise<[boolean, string|null]>} Tuple of (success, error_message)
 */
async function deleteUser(userId) {
  const user = await User.findByPk(userId);
  
  if (!user) {
    return [false, 'User not found'];
  }

  try {
    await user.destroy();
    return [true, null];
  } catch (error) {
    console.error('Failed to delete user:', error);
    return [false, 'Failed to delete account'];
  }
}

module.exports = {
  createUser,
  authenticateUser,
  getUserById,
  getUserByEmail,
  getUserByUsername,
  updateUser,
  changePassword,
  deleteUser,
};

File: backend/app/services/account_service.py
"""
Account service for user management operations.
"""
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError
from typing import Optional
from datetime import datetime

from app.models.user import User
from app.services.auth_service import get_password_hash, verify_password


class AccountService:
    """
    Service class for account-related operations.
    """
    
    @staticmethod
    def create_user(
        db: Session,
        email: str,
        username: str,
        password: str,
        riot_id: Optional[str] = None,
        region: Optional[str] = None
    ) -> tuple[Optional[User], Optional[str]]:
        """
        Create a new user account.
        
        Returns:
            Tuple of (User, None) on success or (None, error_message) on failure
        """
        # Check if email already exists
        if db.query(User).filter(User.email == email).first():
            return None, "Email already registered"
        
        # Check if username already exists
        if db.query(User).filter(User.username == username).first():
            return None, "Username already taken"
        
        # Create new user
        hashed_password = get_password_hash(password)
        user = User(
            email=email,
            username=username,
            hashed_password=hashed_password,
            riot_id=riot_id,
            region=region
        )
        
        try:
            db.add(user)
            db.commit()
            db.refresh(user)
            return user, None
        except IntegrityError:
            db.rollback()
            return None, "Failed to create account"
    
    @staticmethod
    def authenticate_user(
        db: Session,
        email: str,
        password: str
    ) -> Optional[User]:
        """
        Authenticate a user by email and password.
        
        Returns:
            User if authentication successful, None otherwise
        """
        user = db.query(User).filter(User.email == email).first()
        
        if not user:
            return None
        
        if not verify_password(password, user.hashed_password):
            return None
        
        # Update last login
        user.last_login = datetime.utcnow()
        db.commit()
        
        return user
    
    @staticmethod
    def get_user_by_id(db: Session, user_id: int) -> Optional[User]:
        """
        Get a user by their ID.
        """
        return db.query(User).filter(User.id == user_id).first()
    
    @staticmethod
    def get_user_by_email(db: Session, email: str) -> Optional[User]:
        """
        Get a user by their email.
        """
        return db.query(User).filter(User.email == email).first()
    
    @staticmethod
    def get_user_by_username(db: Session, username: str) -> Optional[User]:
        """
        Get a user by their username.
        """
        return db.query(User).filter(User.username == username).first()
    
    @staticmethod
    def update_user(
        db: Session,
        user_id: int,
        username: Optional[str] = None,
        riot_id: Optional[str] = None,
        region: Optional[str] = None
    ) -> tuple[Optional[User], Optional[str]]:
        """
        Update user profile information.
        
        Returns:
            Tuple of (User, None) on success or (None, error_message) on failure
        """
        user = db.query(User).filter(User.id == user_id).first()
        
        if not user:
            return None, "User not found"
        
        # Check username uniqueness if changing
        if username and username != user.username:
            if db.query(User).filter(User.username == username).first():
                return None, "Username already taken"
            user.username = username
        
        if riot_id is not None:
            user.riot_id = riot_id
        
        if region is not None:
            user.region = region
        
        try:
            db.commit()
            db.refresh(user)
            return user, None
        except IntegrityError:
            db.rollback()
            return None, "Failed to update profile"
    
    @staticmethod
    def change_password(
        db: Session,
        user_id: int,
        current_password: str,
        new_password: str
    ) -> tuple[bool, Optional[str]]:
        """
        Change a user's password.
        
        Returns:
            Tuple of (True, None) on success or (False, error_message) on failure
        """
        user = db.query(User).filter(User.id == user_id).first()
        
        if not user:
            return False, "User not found"
        
        if not verify_password(current_password, user.hashed_password):
            return False, "Current password is incorrect"
        
        user.hashed_password = get_password_hash(new_password)
        db.commit()
        
        return True, None
    
    @staticmethod
    def delete_user(db: Session, user_id: int) -> tuple[bool, Optional[str]]:
        """
        Delete a user account.
        
        Returns:
            Tuple of (True, None) on success or (False, error_message) on failure
        """
        user = db.query(User).filter(User.id == user_id).first()
        
        if not user:
            return False, "User not found"
        
        db.delete(user)
        db.commit()
        
        return True, None
